samples/bluetooth/direct_test_mode/src/fem/nrf21540.c

```diff
@@ -514,7 +544,207 @@ int nrf21540_antenna_select(enum nrf21540_ant ant)
+	cfg->rx_pin = nrf_uarte_rx_pin_get(uarte);
+	cfg->cts_pin = nrf_uarte_cts_pin_get(uarte);
+	cfg->rts_pin = nrf_uarte_rts_pin_get(uarte);
+	cfg->baudrate = uarte->BAUDRATE;
+	cfg->config = uarte->CONFIG;
+	cfg->rxd_maxcnt = uarte->RXD.MAXCNT;
+	cfg->rxd_ptr = uarte->RXD.PTR;
+	cfg->txd_maxcnt = uarte->TXD.MAXCNT;
+	cfg->txd_ptr = uarte->TXD.PTR;
+	cfg->iten = nrf_uarte_int_enable_check(uarte, 0xFFFFFFFF);
+}
+
+static void uarte_configuration_restore(NRF_UARTE_Type *uarte,
+					const struct nrf_uarte_config *cfg)
+{
+	nrf_uarte_txrx_pins_set(uarte, cfg->tx_pin, cfg->rx_pin);
+	nrf_uarte_hwfc_pins_set(uarte, cfg->rts_pin, cfg->cts_pin);
+	uarte->CONFIG = cfg->config;
+	nrf_uarte_baudrate_set(uarte, cfg->baudrate);
+	nrf_uarte_rx_buffer_set(uarte, (uint8_t *) cfg->rxd_ptr,
+				cfg->rxd_maxcnt);
+	nrf_uarte_tx_buffer_set(uarte, (uint8_t *) cfg->txd_ptr,
+				cfg->txd_maxcnt);
+	nrf_uarte_int_enable(NRF_UARTE0, cfg->iten);
+}
+
+static int spim_gain_transfer(uint8_t gain)
+{
+	int err;
+	nrfx_err_t nrfx_err;
+	uint8_t buf[NRF21540_SPI_LENGTH_BYTES];
+	nrfx_spim_config_t spim_config = NRFX_SPIM_DEFAULT_CONFIG(
+					DT_PROP(NRF21540_SPI_BUS, sck_pin),
+					DT_PROP(NRF21540_SPI_BUS, mosi_pin),
+					DT_PROP(NRF21540_SPI_BUS, miso_pin),
+					NRFX_SPIM_PIN_NOT_USED);
+
+	spim_config.frequency =
+		get_nrf_spim_frequency(nrf21540_cfg.spi_cfg.frequency);
+
+	err = gpio_pin_set(nrf21540_cfg.spi_cs.gpio_dev,
+			   nrf21540_cfg.spi_cs.gpio_pin, 0);
+	if (err) {
+		return err;
+	}
+
+	nrfx_err = nrfx_spim_init(&spim, &spim_config, NULL, NULL);
+	if (nrfx_err != NRFX_SUCCESS) {
+		return -ENXIO;
+	}
+
+	buf[NRF21540_SPI_COMMAND_ADDR_BYTE] =
+		NRF21540_SPI_WRITE(NRF21540_REG_CONFREG0);
+	buf[NRF21540_SPI_DATA_BYTE] =
+		(gain << NRF21540_BITS_CONFREG0_TX_GAIN_Pos) &
+		NRF21540_BITS_CONFREG0_TX_GAIN_Msk;
+
+	/* Prepare transfer */
+	nrfx_spim_xfer_desc_t xfer_desc = NRFX_SPIM_XFER_TX(buf, sizeof(buf));
+
+	err = gpio_pin_set(nrf21540_cfg.spi_cs.gpio_dev,
+			   nrf21540_cfg.spi_cs.gpio_pin, 1);
+	if (err) {
+		return err;
+	}
+
+	nrfx_err = nrfx_spim_xfer(&spim, &xfer_desc,
+				  NRFX_SPIM_FLAG_NO_XFER_EVT_HANDLER);
+	if (nrfx_err != NRFX_SUCCESS) {
+		return -ENXIO;
+	}
+
+	nrfx_spim_uninit(&spim);
+	spim_events_clear(spim.p_reg);
+
+	return 0;
+}
+
+int nrf21540_tx_gain_set(uint8_t gain)
+{
+	int err;
+	NRF_UARTE_Type *uarte_inst = NRF_UARTE0;
+	const struct device *uart;
+	struct nrf_uarte_config uarte_cfg;
+	bool uart_ready = false;
+	bool uart_irq = false;
+	unsigned int key;
+
+	if (gain > NRF21540_TX_GAIN_Max) {
+		return -EINVAL;
+	}
+
+	if (!atomic_cas(&state, NRF21540_STATE_READY, NRF21540_STATE_BUSY)) {
+		return -EBUSY;
+	}
+
+	memset(&uarte_cfg, 0, sizeof(uarte_cfg));
+
+	key = irq_lock();
+
+	/* Disable UART_0 */
+	uart = device_get_binding("UART_0");
+	if (uart) {
+		uart_ready = device_is_ready(uart);
+	}
+
+	uart_irq = irq_is_enabled(nrfx_get_irq_number(uarte_inst));
+	if (uart_irq) {
+		irq_disable(nrfx_get_irq_number(uarte_inst));
+	}
```

Comment:
Is this actually needed when interrupts are locked?

Reply:
Maybe not. I disable the interrupt because I initialize the SPI which uses the same NVIC interrupt line as the UART so It seems to be more safely for me even if it is redundant.